apps/builder/app/builder/features/ai/ai-command-bar.tsx

```diff
@@ -26,57 +26,49 @@ import {
   StopIcon,
   LargeXIcon,
 } from "@webstudio-is/icons";
-import {
-  useRef,
-  useState,
-  type MouseEvent,
-  type ComponentPropsWithoutRef,
-} from "react";
+import { useRef, useState } from "react";
 import { $isAiCommandBarVisible } from "~/shared/nano-states";
 import { useMediaRecorder } from "./hooks/media-recorder";
 import { useLongPressToggle } from "./hooks/long-press-toggle";
 import { AiCommandBarButton } from "./ai-button";
 import { fetchTranscription } from "./ai-fetch-transcription";
 import { fetchResult } from "./ai-fetch-result";
 
-type PartialButtonProps<T = ComponentPropsWithoutRef<typeof Button>> = {
-  [key in keyof T]?: T[key];
-};
+type Status = "idle" | "recording" | "transcribing" | "ai";
 
 export const AiCommandBar = () => {
-  const [value, setValue] = useState("");
-  const [prompts, setPrompts] = useState<string[]>([]);
-  const [open, setOpen] = useState(false);
-  const [isAudioTranscribing, setIsAudioTranscribing] = useState(false);
-  const [isAiRequesting, setIsAiRequesting] = useState(false);
-  const abortController = useRef<AbortController>();
   const isAiCommandBarVisible = useStore($isAiCommandBarVisible);
+  const [open, setOpen] = useState<boolean>(false);
+
+  const [status, setStatus] = useState<Status>("idle");
+  const abortController = useRef<AbortController>();
+
+  const [prompt, setPrompt] = useState<string>("");
+  const [promptsHistory, setPromptsHistory] = useState<string[]>([]);
+
   const recordButtonRef = useRef<HTMLButtonElement>(null);
-  const guardIdRef = useRef(0);
   const { enableCanvasPointerEvents, disableCanvasPointerEvents } =
     useDisableCanvasPointerEvents();
 
-  const {
-    start,
-    stop,
-    cancel,
-    state: mediaRecorderState,
-  } = useMediaRecorder({
+  const { start, stop, cancel } = useMediaRecorder({
     onComplete: async (file) => {
-      setIsAudioTranscribing(true);
-      guardIdRef.current++;
-      const guardId = guardIdRef.current;
-      const text = await fetchTranscription(file);
-      if (guardId !== guardIdRef.current) {
-        return;
+      setStatus("transcribing");
+      abortController.current = new AbortController();
```

Comment:
This is so so code for async env. You are trying to use the fact that `onComplete` will never be called again until all async code below will be completed. It's not true. 
`onCompleted` doesn't give this guarantee, see `guardIdRef.current++` is not for fun. 
Even every button below is disabled "immediately" to prevent parallel execution of anything, it's React `setState` is not synchronous, there is not guarantee then button couldn't be clicked again before disabled state.

`abortController.current = new AbortController();` everytime you do this having assumptions above you need to check that `abortController.current` is disposed